tpe/plugin/src/KinematicsFeatures.cc

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright (C) 2020 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+#include <ignition/common/Console.hh>
+#include <ignition/math/eigen3/Conversions.hh>
+
+#include "KinematicsFeatures.hh"
+
+using namespace ignition;
+using namespace physics;
+using namespace tpeplugin;
+
+/////////////////////////////////////////////////
+FrameData3d KinematicsFeatures::FrameDataRelativeToWorld(
+  const FrameID &_id) const
+{
+  FrameData3d data;
+
+  // The feature system should never send us the world ID.
+  if (_id.IsWorld())
+  {
+    ignerr << "Given a FrameID belonging to the world. This should not be "
+           << "possible! Please report this bug!\n";
+    assert(false);
+    return data;
+  }
+  // check if id is present and skip if not
+  auto it = this->childIdToParentId.find(_id.ID());
+  if (it == this->childIdToParentId.end())
+  {
+    ignwarn << "Link [" << _id.ID() << "]  is not found." << std::endl;
+    return data;
+  }
+  std::size_t modelId = it->second;
+  auto modelIt = this->models.find(modelId);
+  if (modelIt == this->models.end())
+  {
+    ignwarn << "Parent model ["
+      << modelId
+      << "] of link ["
+      << _id.ID()
+      << "] is not found."
+      << std::endl;
+    return data;
+  }
+  tpelib::Model *model = modelIt->second->model;
+
+  data.pose = math::eigen3::convert(model->GetPose());
+  data.linearVelocity = math::eigen3::convert(model->GetLinearVelocity());
+  data.angularVelocity = math::eigen3::convert(model->GetAngularVelocity());
```

Comment:
I think the assumption here is the pose of link and collision is always consistent with model and by setting the velocity and pose of a model, we set them for all entities inside the model. Tbh I'd rather implement the offset in a separate PR if that's deemed necessary/need arises since this PR is already quite large.

Reply:
Ok, we can leave this for later since this PR is quite large already. My worry is that we can easily forget to come back to this and end up having to hunt this problem in the future. I think any multi-link model would be broken with the current implementation, because I believe such models will always have link offsets.

How about:

* Writing a TODO saying that the link pose needs to be considered
* Ticketing an issue in case we're not planning to tackle this next
* Bonus: detecting when the link pose is not zero and printing a warning (at this point, it may be just a bit more effort to add the link offset to these results)

Reply:
Cool, I have added a TODO https://github.com/ignitionrobotics/ign-physics/pull/46/commits/5215a743061e5a6a54dc193726a50f5cd1ff9480 and ticketed an issue https://github.com/ignitionrobotics/ign-physics/issues/57